### PR TITLE
DateTime.iso8601 fails with an error if a second fraction is present

### LINKED ISSUE #2883: DateTime.iso8601 fails with an error (incompatibility)
```
jruby-1.7.19 :004 > require 'date'
 => true 
jruby-1.7.19 :005 > DateTime.parse '2014-07-08T17:51:36.000Z'
 => #<DateTime: 2014-07-08T17:51:36+00:00 ((2456847j,64296s,0n),+0s,2299161j)> 
jruby-1.7.19 :006 > puts DateTime.parse '2014-07-08T17:51:36.000Z'
2014-07-08T17:51:36+00:00
 => nil 
jruby-1.7.19 :007 > DateTime.iso8601('2014-07-08T17:51:36.000Z')
NoMethodError: undefined method `/' for "000":String
	from /opt/local/rvm/rubies/jruby-1.7.19/lib/ruby/1.9/date.rb:1778:in `new_by_frags'
	from /opt/local/rvm/rubies/jruby-1.7.19/lib/ruby/1.9/date.rb:1829:in `iso8601'
	from (irb):7:in `evaluate'
	from org/jruby/RubyKernel.java:1107:in `eval'
	from org/jruby/RubyKernel.java:1507:in `loop'
	from org/jruby/RubyKernel.java:1270:in `catch'
	from org/jruby/RubyKernel.java:1270:in `catch'
	from /opt/local/rvm/rubies/jruby-1.7.19/bin/irb:13:in `(root)'
```

expected :

```
1.9.3-p551 :001 > require 'date'
 => true 
1.9.3-p551 :002 > DateTime.iso8601('2014-07-08T17:51:36.000Z')
 => #<DateTime: 2014-07-08T17:51:36+00:00 ((2456847j,64296s,0n),+0s,2299161j)> 
1.9.3-p551 :003 > puts DateTime.iso8601('2014-07-08T17:51:36.000Z')
2014-07-08T17:51:36+00:00
 => nil 
```

taken from: http://stackoverflow.com/questions/29786162/datetime-iso86012014-07-08t175136-000z-results-in-nomethoderror

### PR DESCRIPTION
**Pull request by azolotko**
Fixes #2883 

I would be happy to add a test, if I only knew where to put it.

Thanks!

### CODE REVIEW

**Comment by BanzaiMan** [comment]
Keep in mind that `lib/ruby/1.9` is where MRI standard libraries live. Things should not diverge there.

**Comment by azolotko** [comment]
@BanzaiMan Yes, absolutely. Please correct me if I'm wrong, but MRI seems to do a similar thing: https://github.com/ruby/ruby/blob/trunk/ext/date/date_parse.c#L2363

**Comment by kares** [comment]
@BanzaiMan in this case it is acceptable as *date.rb* has gone native somewhere along MRI 1.9.3
... simply no longer there at https://github.com/ruby/ruby/tree/ruby_1_9_3/lib (exists in **ruby_1_9_2**)

\+ this is a follow-up on a previous *date/format.rb* patch: https://github.com/jruby/jruby/commit/b2e1dc1338e88d24c74ffb563b0775bee865526a?diff=unified

@azolotko test/spec will be fine under [regressions](https://github.com/jruby/jruby/tree/jruby-1_7/spec/regression) ... just be aware to [filter it out in 1.8 mode](https://github.com/jruby/jruby/commit/3ac859f61eaaf8a19c636fa58e8df148b6c75418) (still tested against under **jruby-1_7**)

**Comment by azolotko** [comment]
There seems to be some difference in meaning of `:sec_fraction` between `_parse(str, comp)` and `_iso8601(str)`

**Comment by eregon** [comment]
Having the spec in spec/ruby (RubySpec) would be useful as well. 
If you do that, please make a separate commit for the specs.

File: lib/ruby/1.9/date/format.rb [hunk 1]
```diff
@@ -937,7 +937,7 @@ def self._iso8601(str) # :nodoc:
         h[:sec] = i sec if sec
       end
 
-      h[:sec_fraction] = sec_fraction if sec_fraction
+      h[:sec_fraction] = Rational(sec_fraction.to_i, 10**sec_fraction.size) if sec_fraction
```
Comment (eregon) [inline lib/ruby/1.9/date/format.rb]: What about `Float("0#{sec_fraction}")` ?

Comment (kares) [inline lib/ruby/1.9/date/format.rb]: cause of MRI compatibility: 
```
1.9.3-p551 :002 > date = DateTime.iso8601('2014-07-08T17:51:36.013Z')
 => #<DateTime: 2014-07-08T17:51:36+00:00 ((2456847j,64296s,13000000n),+0s,2299161j)> 
1.9.3-p551 :003 > date.sec_fraction
 => (13/1000) 
1.9.3-p551 :004 > date.sec_fraction.class
 => Rational 
1.9.3-p551 :005 > date.second_fraction
 => (13/1000) 
```

**Comment by kares** [comment]
cherry-picked commit on **jruby-1_7** as https://github.com/jruby/jruby/commit/4e822a112cfec7e06f4e248e74cc1ca2a549d109 ... great work Alex, esp. that you matched MRI ... thanks.

**Pull request closed without merging**